### ISSUE TITLE
WebAssembly implementation of Zarr

### ISSUE THREAD

**Issue by jakirkham** [zarr-developers/community#14]
Would be really nice to have a [WebAssembly]( https://webassembly.org/ ) implementation of Zarr. This could make it possible to load Zarr files in the browser for viewing or for computation. Could also be useful to be able to work with in-memory Zarr objects in the browser. Given there has already been some good work to get [Python and NumPy]( http://droettboom.com/blog/2018/04/04/python-in-the-browser/ ) into the browser using Emscripten, it may be possible to just run the Python Zarr implementation in the browser. Though other compression algorithms not in the standard Python library will require getting Blosc and Numcodecs into WebAssembly. ( https://github.com/Blosc/c-blosc/issues/238 )

Note: WebAssembly is [well supported]( https://caniuse.com/wasm ). For older browsers one can [convert WebAssembly to asm.js]( https://github.com/WebAssembly/binaryen/blob/master/src/wasm2asm.h ), which is [pretty well supported]( https://caniuse.com/asmjs ). In the worst case, asm.js is still valid JavaScript. So can be run as JavaScript (albeit slowly).

**Comment by dhirschfeld** [comment]
Related: https://gitter.im/zarr-developers/community?at=5ddc7f1c55bbed7ade461091
https://github.com/gzuidhof/zarr.js

Not WASM but aims to make `zarr` files accessible from the browser

**Comment by jakirkham** [comment]
Also noting there is a Scala implementation that can be compiled to JavaScript. Relevant discussion and more info in issue ( https://github.com/zarr-developers/community/issues/15 ). That said, I'm not aware of a good path from Scala to WebAssembly.

**Comment by vdwees** [comment]
I'd be interested in contributing to a rust and wasm implementation of zarr. Would anyone like to collaborate on this?

**Comment by thewtex** [comment]
A wasm implementation of Zarr-Blosc decompression is here:

https://github.com/Kitware/itk-vtk-viewer/blob/7f82bbff02b6e8d847c76457fc07979be07c7ad5/src/bloscZarrDecompress.js

If there is interest, this could be separated out into a new package, and the corresponding compression function added (it has already been implemented in C/Emscripten). It would make sense to use this as the decompression for JavaScript / Typescript libraries like @gzuidhof 's zarr.js or  @freeman-lab 's zarr-js.

This implementation supports all blosc codec's. It also uses a pool of web workers to decompress a set of chunks in parallel and optimize wasm compilation.

Here is what it looks like in action:

https://kitware.github.io/itk-vtk-viewer/app/?fileToLoad=https://thewtex.github.io/allen-ccf-itk-vtk-zarr/average_template_50_chunked.zarr

**Comment by alimanfoo** [comment]
> Here is what it looks like in action:
> 
> https://kitware.github.io/itk-vtk-viewer/app/?fileToLoad=https://thewtex.github.io/allen-ccf-itk-vtk-zarr/average_template_50_chunked.zarr

Wow, that is mega cool. What is it?

**Comment by vdwees** [comment]
@thewtex It would be awesome to make it a separate package.  I am new to WebAssembly, but I'd be happy to contribute where I can.

**Comment by thewtex** [comment]
> Wow, that is mega cool. What is it?

A brain atlas averaged from 1675 mice :mouse: :mouse: :mouse: 

> @thewtex It would be awesome to make it a separate package. I am new to WebAssembly, but I'd be happy to contribute where I can.

Great, we'll create a package, your help is appreciated.

**Comment by manzt** [comment]
I just made [`numcodec.js`](https://github.com/manzt/numcodecs.js) public which has a WASM blosc codec. Hopefully this will help others use blosc in their applications!

**Comment by jakirkham** [comment]
> I just made [`numcodec.js`](https://github.com/manzt/numcodecs.js) public which has a WASM blosc codec. Hopefully this will help others use blosc in their applications!

Nice work! Thanks for sharing @manzt. I wonder how hard it is to get Zarr usable from WebAssembly then (as it is pure Python at that point)

cc @rth @mdboom (who may be interested ;)

**Comment by rth** [comment]
> I wonder how hard it is to get Zarr usable from WebAssembly then (as it is pure Python at that point)

If it is pure python (and has pure python wheels) you could install it from PyPi with pyodide, but you would still need to write some code to interact with those JS/WASM libraries where currently it uses other Python package with C-extensions..

**Comment by Farkal** [comment]
Hey there, i am very interested by the zarr format so i am available to create a WebAssembly/Rust lib but i would like to directly implement the v3 spec. After reading some of the differents topic of the zarr spec repo the spec for the v3 seems pretty great. Do you think i could start some implementation ? Or should i wait for the python impl first ?

**Comment by jakirkham** [comment]
Hey @Farkal, welcome! That sounds great! 😄

Would be nice to have other people trying out the spec in other languages. This can help inform whether what we have in the spec makes sense or if it needs further modification. FWIW there is a WIP Python implementation here ( https://github.com/alimanfoo/zarrita ). Also we have been engaging with some folks from [QuantStack]( https://quantstack.net ) on the C++ side and with [NetCDF]( https://www.unidata.ucar.edu/software/netcdf/ ) on the C side. It would be really interesting to see whether things makes sense from the WebAssembly/Rust side.

Also we have a weekly spec meeting details in issue ( https://github.com/zarr-developers/community/issues/33 ) if you would be interested in stopping by. Would be nice to say hi and learn a bit more about what you are working as well as how we can help 🙂

**Comment by oeway** [comment]
FYI: `zarr-python` and `numcodecs` are compiled into WASM and available as modules in Pyodide after [this PR](https://github.com/iodide-project/pyodide/pull/938).

[Click here to try a live demo](https://notebook.imjoy.io/?ipynb=%257B%2522path%2522%253A%2522Untitled.ipynb%2522%252C%2522name%2522%253A%2522Untitled.ipynb%2522%252C%2522content%2522%253A%257B%2522cells%2522%253A%255B%257B%2522metadata%2522%253A%257B%2522trusted%2522%253Atrue%257D%252C%2522cell_type%2522%253A%2522code%2522%252C%2522source%2522%253A%2522import%2520zarr%255Cnz%2520%253D%2520zarr.zeros%2528%252810000%252C%252010000%2529%252C%2520chunks%253D%25281000%252C%25201000%2529%252C%2520dtype%253D%27i4%27%2529%255Cnz%2522%252C%2522execution_count%2522%253A1%252C%2522outputs%2522%253A%255B%257B%2522output_type%2522%253A%2522stream%2522%252C%2522text%2522%253A%2522%252Flib%252Fpython3.8%252Fsite-packages%252Fnumpy%252Fma%252Fcore.py%253A4467%253A%2520SyntaxWarning%253A%2520%255C%2522is%255C%2522%2520with%2520a%2520literal.%2520Did%2520you%2520mean%2520%255C%2522%253D%253D%255C%2522%253F%255Cn%2520%2520if%2520self.shape%2520is%2520%2528%2529%253A%255Cn%2522%252C%2522name%2522%253A%2522stderr%2522%257D%252C%257B%2522output_type%2522%253A%2522execute_result%2522%252C%2522execution_count%2522%253A1%252C%2522data%2522%253A%257B%2522text%252Fplain%2522%253A%2522%253Czarr.core.Array%2520%252810000%252C%252010000%2529%2520int32%253E%2522%257D%252C%2522metadata%2522%253A%257B%257D%257D%255D%257D%252C%257B%2522metadata%2522%253A%257B%2522trusted%2522%253Atrue%257D%252C%2522cell_type%2522%253A%2522code%2522%252C%2522source%2522%253A%2522%2522%252C%2522execution_count%2522%253Anull%252C%2522outputs%2522%253A%255B%255D%257D%255D%252C%2522metadata%2522%253A%257B%2522kernelspec%2522%253A%257B%2522name%2522%253A%2522python3%2522%252C%2522display_name%2522%253A%2522Python%25203%2522%252C%2522language%2522%253A%2522python%2522%257D%257D%252C%2522nbformat%2522%253A4%252C%2522nbformat_minor%2522%253A2%257D%257D) with Pyodide + zarr running completely in the browser.
(Only works with Chrome or FireFox)

**Comment by jakirkham** [comment]
Very cool! Thanks for sharing Wei 😄

cc @martindurant (who may be interested in `fsspec` usage)

**Comment by martindurant** [comment]
This was discussed a bit on gitter. fsspec for Pyodide seems like a big benefit with or without zarr, but indeed the sync/thread stuff adds complexity that in this environment I don't think I'm in the best place to tackle. Happy to help test, though!

**Comment by jakirkham** [comment]
It looks like Pyodide is including Zarr & Numcodecs, which is cool to see 😄

**Comment by joshmoore** [comment]
@jakirkham : where does that leave this issue? :)

**Comment by oeway** [comment]
> It looks like Pyodide is including Zarr & Numcodecs, which is cool to see 😄

I added that two libraries to pyodide a while ago, it works with in-memory data but still very limited for any real application because we cannot support remote storage backends. 

Not sure if this is discussed already in the zarr community, but the key feature to make that work is to support async store (with asyncio). The native python implementation of fsspec uses threading to convert async calls into sync, but multi-threading in pyodide is not supported yet, it will only work if zarr supports async store (meaning the __getitem__ function will be `async`).

**Comment by martindurant** [comment]
Working on async zarr at https://github.com/martindurant/async-zarr as part of a company hack week

**Comment by martindurant** [comment]
Already works in normal python asyncio, and maybe works in pyscript too, just need to write some HTML or something...

**Comment by MSanKeys963** [comment]
> Working on async zarr at https://github.com/martindurant/async-zarr as part of a company hack week

Thanks for sharing the details, @martindurant. May I know the exact dates for the hack week?
I'd like to post publicly about this to invite more contributors.

**Comment by jakirkham** [comment]
IIUC it is a hack week Anaconda is running for its employees

**Comment by martindurant** [comment]
That's correct; and the hack is now over.
I made this little video of the current state: https://drive.google.com/file/d/1Ll-Lr_3Ckf_-WIlBkIPx4H8Kmz9lz4b9/view?usp=sharing

**Comment by MSanKeys963** [comment]
Thanks a lot, @martindurant. This is great.
I'll share this across our social media to look at so that we can get the word out and invite new users/contributors.
